runtime/scripts/math.js

```diff
@@ -1972,6 +1972,28 @@ var math = Numbas.math = /** @lends Numbas.math */ {
             product = math.mul(product, list[i]);
         }
         return product;
+    },
+    /** Weighted random given a list of [item, probability] 
+     * 
+     * @param {Array} list
+     * @returns {number}
+    */
+     weighted_random: function(list) {
+        var item;
+        var prob;
+        weights = new Map();
+        var probSum = 0;
+        for (var i = 0; i < list.length; i++) {
+            [item, prob] = list[i];
+            probSum += prob;
+            weights.set(probSum, item);
+        }
+        var target = Math.floor(Math.random() * probSum);
```

Comment:
I don't know why `Math.floor` is used here. I think it's incorrect: for example, `weighted_random([ [1, 0.1], [2,0.2] ])` will only ever return `1` as this code stands.